src/devices/Mcp3428/Mcp342x.cs

```diff
@@ -0,0 +1,387 @@
+        /// </exception>
+        /// <autogeneratedoc />
+        protected void OneShotRead(int channel = -1)
+        {
+            if (Mode != ModeEnum.OneShot)
+                throw new IOException("Device is not in One-Shot mode");
+
+            _isReadyBit = false;
+            var conf = Helpers.SetReadyBit(LastConfigByte, false);
+            if (channel >= 0 && channel != LastChannel)
+            {
+                conf = Helpers.SetChannelBits(conf, channel);
+            }
+
+            WriteConfig(conf);
+            using (var source = new CancellationTokenSource(TimeSpan.FromMilliseconds(WaitTime * 5)))
+            {
+                WaitForConversion(TimeSpan.FromMilliseconds(WaitTime), cancellationToken: source.Token);
+
+                if (!_isReadyBit)
+                {
+                    throw new IOException($"ADC Conversion was not ready after {WaitTime * 5} ms.");
+                }
+            }
+        }
+
+        protected void WaitForConversion(TimeSpan? waitSpan = null, Action<int> progressCallback = null, CancellationToken cancellationToken = default)
+        {
+            waitSpan = waitSpan ?? TimeSpan.FromMilliseconds(WaitTime);
+            var allms = 0;
+            _isReadyBit = false;
+            while (!_isReadyBit && !cancellationToken.IsCancellationRequested)
+            {
+                _i2cDevice.Read(_readBuffer);
+                ReadConfigByte(LastConfigByte);
+                if (_isReadyBit)
+                    break;
+
+                Thread.Sleep(waitSpan.Value);
+                cancellationToken.ThrowIfCancellationRequested();
+                allms += (int)(waitSpan.Value.TotalMilliseconds);
+                progressCallback?.Invoke(allms);
+            }
+            cancellationToken.ThrowIfCancellationRequested();
+        }
+
+        protected void ReadConfigByte(byte config)
+        {
+            _isReadyBit = (config & Helpers.Masks.ReadyMask) == 0; // Negated bit
+            _lastChannel = (byte)((config & Helpers.Masks.ChannelMask) >> 5);
+            _mode = (ModeEnum)(config & Helpers.Masks.ModeMask);
+            _pgaGain = (GainEnum)(config & Helpers.Masks.GainMask);
+            _resolution = (ResolutionEnum)(config & Helpers.Masks.ResolutionMask);
+        }
+
+        protected double ReadValue(int channel = -1)
+        {
+            if (Mode == ModeEnum.OneShot)
+            {
+                OneShotRead(channel);
+            }
+            else
+            {
+                if (channel > 0 && channel != LastChannel)
+                {
+                    var conf = Helpers.SetChannelBits(LastConfigByte, channel);
+                    WriteConfig(conf);
+                }
+
+                using (var source = new CancellationTokenSource(TimeSpan.FromMilliseconds(WaitTime * 5)))
+                {
+                    WaitForConversion(TimeSpan.FromMilliseconds(WaitTime / 5), cancellationToken: source.Token);
```

Comment:
why is WaitTime divided by 5? this means the token is valid 25x longer, is this intentional?

Reply:
The `WaitTime / 5` is for when the start of the reading is not in sync with the conversion.  So I poll it multiple times in a period because it may have completed. `WaitForConversion` Actually waits forever if not cancelled, but after every `waitSpan` (input) time it polls the device.

The code actually means it tries to read five times in every nominal conversion period. And if it does not complete after five nominal periods the read gets cancelled.